[70] Add getters for the processed AI and Genomics data

Jack-0-0 opened the pull request:
Closes #70.

This PR adds getters for the processed AI and Genomics GtR, Crunchbase and OpenAlex data. 
 
---

Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [ ] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained the feature in this PR or (better) in `output/reports/`
- [x] I have requested a code review

Commit:
Add getter for gtr data

Commit:
Fix error message

Commit:
Shorten error message

Commit:
Update crunchbase_definitions to save cb_org_ids

Commit:
Add getter for ai genomics crunchbase org ids

Commit:
Add openalex getters

Commit:
Remove openalex abstracts getter

Commit:
Standardise error messages

Commit:
Add note about getters to readmes

georgerichardson commented on ai_genomics/getters/crunchbase.py:
```diff
@@ -0,0 +1,17 @@
+            PROJECT_DIR / "inputs/data/crunchbase/ai_genomics_org_ids.csv",
+            index_col=0,
+            names=["cb_org_id"],
+            skiprows=[0],
```
Is there a reason why this file can't be formatted to remove this arg?

georgerichardson commented on ai_genomics/getters/crunchbase.py:
```diff
@@ -0,0 +1,17 @@
+            names=["cb_org_id"],
+            skiprows=[0],
+        )
+    except FileNotFoundError:
+        logger.error(
+            "FileNotFoundError: To create the missing file, run ai_genomics/analysis/crunchbase_definitions.py"
+        )
```
Only logging an error means that the code will continue to execute after this function is called regardless of the outcome, so it will fail relatively quietly. In this case is that the user might only be alerted when they try to perform an operation on the data and trigger a `NoneType` error.

While logging has the advantage of providing useful information in the console and the ability to send logs to a logfile, raising an exception is more suitable for controlling the flow of code. There's no reason you can't have both though (more detailed explanations [here](https://stackoverflow.com/questions/51206787/between-raise-error-and-logging-the-error-which-one-is-a-better-practice-in-pyt)).

Jack-0-0 replied:
Good point, I'll update it so that there is an error logged and the exception is raised.

georgerichardson commented on ai_genomics/getters/gtr.py:
```diff
@@ -0,0 +1,24 @@
+    """Get AI and Genomics GtR data
+
+    Args:
+        data_type: "projects" or "publications"
```
maybe it would make more sense to call this `entity_type` or something similar? When talking about data types I would normally think of str, int, etc.

georgerichardson replied:
Also a nice to have would be some handling to raise an error if another option (or typo) is passed in.

georgerichardson commented on ai_genomics/pipeline/README.md:
```diff
@@ -29,10 +29,18 @@ aws s3 cp s3://ai-genomics/inputs/openalex/concepts.json inputs/openalex/concept
 To fetch other article category data we use to label the OpenAlex data.
 
-Run `python ai_genomics/analysis/openalex_definition.py` to implement the definitions and generate results. This includes printing stats in the console and saving relevant tables in `inputs/data`. This will also save figures to `outputs/figures`, the provisional ai genomics OpenAlex dataset to `outputs/ai_genomics_provisional_dataset.csv` and examples from the dataset to `outputs/openalex_example ...
+Run `python ai_genomics/analysis/openalex_definition.py` to implement the definitions and generate results. This includes printing stats in the console and saving relevant tables in `inputs/data`. This will also save figures to `outputs/figures`, the provisional ai genomics OpenAlex dataset to `outputs/ai_genomics_provisional_dataset.csv` and examples from the dataset to `outputs/openalex_example ...
```
I know this is just a typo fix, but do you know what is meant by "implement the definitions" here?

Jack-0-0 replied:
I think Juan means implementing the definition of AI and the definition of genomics to find the combined AI and genomics works.

georgerichardson commented on ai_genomics/pipeline/README.md:
```diff
@@ -29,10 +29,18 @@ aws s3 cp s3://ai-genomics/inputs/openalex/concepts.json inputs/openalex/concept
-Run `python ai_genomics/analysis/openalex_definition.py` to implement the definitions and generate results. This includes printing stats in the console and saving relevant tables in `inputs/data`. This will also save figures to `outputs/figures`, the provisional ai genomics OpenAlex dataset to `outputs/ai_genomics_provisional_dataset.csv` and examples from the dataset to `outputs/openalex_example ...
+Run `python ai_genomics/analysis/openalex_definition.py` to implement the definitions and generate results. This includes printing stats in the console and saving relevant tables in `inputs/data`. This will also save figures to `outputs/figures`, the provisional ai genomics OpenAlex dataset to `outputs/ai_genomics_provisional_dataset.csv` and examples from the dataset to `outputs/openalex_example ...
+
+The AI and genomics OpenAlex works can be loaded using `ai_genomics.getters.openalex.get_openalex_ai_genomics_works`
```
```suggestion
The AI and genomics OpenAlex works can be loaded using `ai_genomics.getters.openalex.get_openalex_ai_genomics_works`.
```

georgerichardson commented on ai_genomics/getters/gtr.py:
```diff
@@ -0,0 +1,24 @@
+    try:
+        return pd.read_json(
+            GTR_INPUTS_DIR / f"gtr_ai_genomics_{data_type}.json"
+        ).rename(columns=lambda x: re.sub(r"(?<!^)(?=[A-Z])", "_", x).lower())
```
Can this be moved into the pipeline that produces the data?

georgerichardson commented on ai_genomics/getters/gtr.py:
```diff
@@ -0,0 +1,24 @@
+            GTR_INPUTS_DIR / f"gtr_ai_genomics_{data_type}.json"
+        ).rename(columns=lambda x: re.sub(r"(?<!^)(?=[A-Z])", "_", x).lower())
+    except ValueError:
+        logger.error(
+            "ValueError: To create the missing file, run ai_genomics/analysis/gtr_definitions.py"
```
Same as above

georgerichardson commented on ai_genomics/getters/openalex.py:
```diff
@@ -136,3 +137,15 @@ def work_abstracts(discipline: str, years: List) -> Dict:
+        return pd.read_csv(
+            f"{PROJECT_DIR}/outputs/ai_genomics_provisional_dataset.csv", index_col=0
+        ).drop(columns=["index"])
+    except FileNotFoundError:
+        logger.error(
+            "FileNotFoundError: To create the missing file, run ai_genomics/analysis/openalex_definition.py"
+        )
```
Same as above

georgerichardson commented on ai_genomics/getters/openalex.py:
```diff
@@ -136,3 +137,15 @@ def work_abstracts(discipline: str, years: List) -> Dict:
+    try:
+        return pd.read_csv(
+            f"{PROJECT_DIR}/outputs/ai_genomics_provisional_dataset.csv", index_col=0
+        ).drop(columns=["index"])
```
If `index` will always be dropped then best to incorporate this into the pipeline that produces the data too.

georgerichardson reviewed (commented):
Thanks for making these getters! They are clear and consistent :)

Have made a few minor edits, but the main comments relate to:
1. Checking whether you think this is the right error handling strategy for the getters
2. Considering whether the getters are the right place for the specific data manipulations that are included, rather than the pipelines that produce those data.

Commit:
Update getters to raise exceptions on missing file

Commit:
Update gtr getter variable data_type to entity_type and add ValueError

Commit:
Add missing .

Commit:
Move crunchbase getter processing to creation file

Commit:
Move openalex getter processing to creation file

Commit:
Move gtr getter processing to creation file

Jack-0-0 commented:
> Thanks for making these getters! They are clear and consistent :)
> 
> Have made a few minor edits, but the main comments relate to:
> 
> 1. Checking whether you think this is the right error handling strategy for the getters
> 2. Considering whether the getters are the right place for the specific data manipulations that are included, rather than the pipelines that produce those data.

@georgerichardson Thanks for the review, I've updated the error handling and moved the processing out of the getters and into the pipelines 🙂
